tests/record.cpp

```diff
@@ -0,0 +1,553 @@
+    }
+
+    jit_freeze_destroy(recording);
+}
+
+/**
+ * This tests recording and replay of a horizontal reduction operation (hsum).
+ */
+TEST_BOTH(07_reduce_hsum) {
+    Recording *recording;
+
+    jit_log(LogLevel::Info, "Recording:");
+    {
+        UInt32 i0(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+        UInt32 r0 = opaque<UInt32>(45, 1);
+
+        uint32_t inputs[] = {
+            i0.index(),
+        };
+
+        jit_freeze_start(Backend, inputs, 1);
+
+        UInt32 o0 = hsum(i0);
+        o0.schedule();
+        jit_eval();
+
+        uint32_t outputs[] = {
+            o0.index(),
+        };
+
+        recording = jit_freeze_stop(Backend, outputs, 1);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+    }
+
+    jit_log(LogLevel::Info, "Replay:");
+    {
+        UInt32 i0(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        UInt32 r0 = opaque<UInt32>(55, 1);
+
+        uint32_t inputs[] = {
+            i0.index(),
+        };
+        uint32_t outputs[1];
+
+        jit_freeze_replay(recording, inputs, outputs);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+    }
+
+    jit_freeze_destroy(recording);
+}
+
+/**
+ * Tests recording of a prefix sum operation with different inputs at replay.
+ */
+TEST_BOTH(08_prefix_sum) {
+    Recording *recording;
+
+    jit_log(LogLevel::Info, "Recording:");
+    {
+        UInt32 i0(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+        UInt32 r0(0, 1, 3, 6, 10, 15, 21, 28, 36, 45);
+
+        uint32_t inputs[] = {
+            i0.index(),
+        };
+
+        jit_freeze_start(Backend, inputs, 1);
+
+        uint32_t o0 = jit_var_block_prefix_reduce(
+            ReduceOp::Add, i0.index(), jit_var_size(i0.index()), 0, 0);
+        jit_var_schedule(o0);
+        jit_eval();
+
+        uint32_t outputs[] = {
+            o0,
+        };
+
+        recording = jit_freeze_stop(Backend, outputs, 1);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+    }
+
+    jit_log(LogLevel::Info, "Replay:");
+    {
+        UInt32 i0(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        UInt32 r0(1, 3, 6, 10, 15, 21, 28, 36, 45, 55);
+
+        uint32_t inputs[] = {
+            i0.index(),
+        };
+        uint32_t outputs[1];
+
+        jit_freeze_replay(recording, inputs, outputs);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+    }
+
+    jit_freeze_destroy(recording);
+}
+
+/**
+ * Basic addition test.
+ * Supplying a different input should replay the operation, with this input.
+ * In this case, the input at replay is incremented and should result in an
+ * incremented output.
+ */
+TEST_BOTH(9_resized_input) {
```

Comment:
Wouldn't be needed anymore if you change the width during replay for all of the previous tests.